File: yarl-core-java/src/test/java/hr/tjakopan/yarl/PolicyAsyncTest.java
package hr.tjakopan.yarl;

import hr.tjakopan.yarl.PolicyResult.Success;
import hr.tjakopan.yarl.noop.AsyncNoOpPolicy;
import hr.tjakopan.yarl.retry.AsyncRetryPolicy;
import hr.tjakopan.yarl.test.helpers.TestResult;
import org.junit.Test;

import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicReference;

import static hr.tjakopan.yarl.Functions.*;
import static org.assertj.core.api.Assertions.*;

public class PolicyAsyncTest {
  private static final String RESULT_SUCCESS = "Result was a success.";
  private static final String RESULT_FAILURE = "Result was a failure.";
  private static final String RESULT_FAILURE_WITH_RESULT = "Result was a failure with handled result.";
  private static final String RESULT_FAILURE_WITH_EXCEPTION = "Result was a failure with exception.";
  private static final String NULL_PARAMETER = "Parameter specified as non-null is null";

  //<editor-fold desc="execute tests">
  @Test
  public void executingThePolicyFunctionShouldExecuteTheSpecifiedAsyncFunctionAndReturnTheResult() {
    final var policy = AsyncRetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry();

    final var future = policy.executeAsync(() -> CompletableFuture.completedFuture(TestResult.GOOD));

    assertThat(future.join()).isEqualTo(TestResult.GOOD);
  }
  //</editor-fold>

  //<editor-fold desc="executeAndCapture tests">
  @Test
  public void executingThePolicyFunctionSuccessfullyShouldReturnSuccessResult() {
    final var future = AsyncRetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry()
      .executeAndCaptureAsync(() -> CompletableFuture.completedFuture(TestResult.GOOD));

    final var result = future.join();

    result.onSuccess(fromConsumer2((r, context) -> assertThat(r).isEqualTo(TestResult.GOOD)))
      .onFailure(fromConsumer2(((faultType, context) -> fail(RESULT_FAILURE))));
  }

  @Test
  public void executingThePolicyFunctionAndFailingWithAHandledExceptionTypeShouldReturnFailureResultIndicatingThatExceptionTypeIsOneHandledByThisPolicy() {
    final var handledException = new ArithmeticException();
    final var future = AsyncRetryPolicy.<TestResult>builder()
      .handle(ArithmeticException.class)
      .retry()
      .executeAndCaptureAsync(() -> {
        throw handledException;
      });

    final var result = future.join();

    result.onFailureWithException(fromConsumer4(throwable -> exceptionType -> ((faultType, context) -> {
      assertThat(throwable).isSameAs(handledException);
      assertThat(exceptionType).isSameAs(ExceptionType.HANDLED_BY_THIS_POLICY);
      assertThat(faultType).isSameAs(FaultType.EXCEPTION_HANDLED_BY_THIS_POLICY);
    })))
      .onSuccess((integer, context) -> fail(RESULT_SUCCESS))
      .onFailureWithResult((integer, faultType, context) -> fail(RESULT_FAILURE_WITH_RESULT));
  }

  @Test
  public void executingThePolicyFunctionAndFailingWithAnUnhandledExceptionTypeShouldReturnFailureResultIndicatingThatExceptionTypeIsUnhandledByThisPolicy() {
    final var unhandledException = new Exception();
    final var future = AsyncRetryPolicy.builder()
      .handle(ArithmeticException.class)
      .retry()
      .executeAndCaptureAsync(() -> {
        throw new RuntimeException(unhandledException);
      });

    final var result = future.join();

    result.onFailureWithException(fromConsumer4(throwable -> exceptionType -> (faultType, context) -> {
      assertThat(throwable.getCause()).isSameAs(unhandledException);
      assertThat(exceptionType).isSameAs(ExceptionType.UNHANDLED);
      assertThat(faultType).isSameAs(FaultType.UNHANDLED_EXCEPTION);
    }))
      .onSuccess((integer, context) -> fail(RESULT_SUCCESS))
      .onFailureWithResult((integer, faultType, context) -> fail(RESULT_FAILURE_WITH_RESULT));
  }

  @Test
  public void executingThePolicyFunctionAndFailingWithAHandledResultShouldReturnFailureResultIndicatingThatResultIsOneHandledByThisPolicy() {
    final var handledResult = TestResult.FAULT;
    final var future = AsyncRetryPolicy.<TestResult>builder()
      .handleResult(handledResult)
      .retry()
      .executeAndCaptureAsync(() -> CompletableFuture.completedFuture(handledResult));

    final var result = future.join();

    result.onFailureWithResult(fromConsumer3(testResult -> (faultType, context) -> {
      assertThat(testResult).isSameAs(handledResult);
      assertThat(faultType).isSameAs(FaultType.RESULT_HANDLED_BY_THIS_POLICY);
    }))
      .onSuccess((testResult, context) -> fail(RESULT_SUCCESS))
      .onFailureWithException((throwable, exceptionType, faultType, context) -> fail(RESULT_FAILURE_WITH_EXCEPTION));
  }

  @Test
  public void executingThePolicyFunctionAndReturningAnUnhandledResultShouldReturnResultNotIndicatingAnyFailure() {
    final var handledResult = TestResult.FAULT;
    final var unhandledResult = TestResult.GOOD;
    final var future = AsyncRetryPolicy.<TestResult>builder()
      .handleResult(handledResult)
      .retry()
      .executeAndCaptureAsync(() -> CompletableFuture.completedFuture(unhandledResult));

    final var result = future.join();

    assertThat(result.isSuccess()).isTrue();
    //noinspection rawtypes
    assertThat(((Success) result).getResult()).isSameAs(unhandledResult);
  }
  //</editor-fold>

  //<editor-fold desc="context tests">
  @Test
  public void executingThePolicyFunctionShouldThrowWhenContextDataIsNull() {
    //noinspection ConstantConditions
    assertThatExceptionOfType(IllegalArgumentException.class)
      .isThrownBy(() -> AsyncRetryPolicy.builder()
        .handle(ArithmeticException.class)
        .retry()
        .executeAsync((Map<String, Object>) null, context -> CompletableFuture.completedFuture(2)))
      .withMessageContaining(NULL_PARAMETER);
  }

  @Test
  public void executingThePolicyFunctionShouldThrowWhenContextIsNull() {
    //noinspection ConstantConditions
    assertThatExceptionOfType(IllegalArgumentException.class)
      .isThrownBy(() -> AsyncRetryPolicy.builder()
        .handle(ArithmeticException.class)
        .retry()
        .executeAsync((Context) null, context -> CompletableFuture.completedFuture(2)))
      .withMessageContaining(NULL_PARAMETER);
  }

  @Test
  public void executingAndCapturingThePolicyFunctionShouldThrowWhenContextDataIsNull() {
    //noinspection ConstantConditions
    assertThatExceptionOfType(IllegalArgumentException.class)
      .isThrownBy(() -> AsyncRetryPolicy.builder()
        .handle(ArithmeticException.class)
        .retry()
        .executeAndCaptureAsync((Map<String, Object>) null, context -> CompletableFuture.completedFuture(2)))
      .withMessageContaining(NULL_PARAMETER);
  }

  @Test
  public void executingAndCapturingThePolicyFunctionShouldThrowWhenContextIsNull() {
    //noinspection ConstantConditions
    assertThatExceptionOfType(IllegalArgumentException.class)
      .isThrownBy(() -> AsyncRetryPolicy.builder()
        .handle(ArithmeticException.class)
        .retry()
        .executeAndCaptureAsync((Context) null, context -> CompletableFuture.completedFuture(2)))
      .withMessageContaining(NULL_PARAMETER);
  }

  @Test
  public void executingThePolicyFunctionShouldPassContextToExecutedDelegate() {
    final var operationKey = "SomeKey";
    final var executionContext = Context.builder()
      .operationKey(operationKey)
      .build();
    final var capturedContext = new AtomicReference<Context>();

    new AsyncNoOpPolicy<>()
      .executeAsync(executionContext, context -> {
        capturedContext.set(context);
        return CompletableFuture.completedFuture(null);
      })
      .join();

    assertThat(capturedContext.get().getOperationKey()).isEqualTo(operationKey);
  }

  @Test
  public void executingAndCapturingThePolicyFunctionShouldPassContextToExecutedDelegate() {
    final var operationKey = "SomeKey";
    final var executionContext = Context.builder()
      .operationKey(operationKey)
      .build();
    final var capturedContext = new AtomicReference<Context>();

    new AsyncNoOpPolicy<>()
      .executeAndCaptureAsync(executionContext, context -> {
        capturedContext.set(context);
        return CompletableFuture.completedFuture(null);
      })
      .join();

    assertThat(capturedContext.get().getOperationKey()).isEqualTo(operationKey);
  }

  @Test
  public void executingAndCapturingThePolicyFunctionShouldPassContextToPolicyResult() {
    final var operationKey = "SomeKey";
    final var executionContext = Context.builder()
      .operationKey(operationKey)
      .build();
    final var future = new AsyncNoOpPolicy<>()
      .executeAndCaptureAsync(executionContext, context -> null);

    final var result = future.join();

    assertThat(result.getContext().getOperationKey()).isEqualTo(operationKey);
  }
  //</editor-fold>
}

File: yarl-core/src/test/kotlin/hr/tjakopan/yarl/PolicyAsyncTest.kt
package hr.tjakopan.yarl

import hr.tjakopan.yarl.noop.asyncNoOp
import hr.tjakopan.yarl.retry.asyncRetry
import hr.tjakopan.yarl.test.helpers.TestResult
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runBlockingTest
import org.assertj.core.api.Assertions.assertThat
import kotlin.test.Test
import kotlin.test.fail

const val RESULT_SUCCESS = "Result was a success."
const val RESULT_FAILURE = "Result was a failure."
const val RESULT_FAILURE_WITH_RESULT = "Result was a failure with handled result."
const val RESULT_FAILURE_WITH_EXCEPTION = "Result was a failure with exception."

@ExperimentalCoroutinesApi
class PolicyAsyncTest {

  //<editor-fold desc="execute tests">
  @Test
  fun `executing the policy function should execute the specified function and return the result`() = runBlockingTest {
    val policy = Policy.asyncRetry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry()

    val result: TestResult = policy.execute { TestResult.GOOD }

    assertThat(result).isEqualTo(TestResult.GOOD)
  }
  //</editor-fold>

  //<editor-fold desc="executeAndCapture tests">
  @Test
  fun `executing the policy function successfully should return success result`() = runBlockingTest {
    val result = Policy.asyncRetry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry()
      .executeAndCapture { TestResult.GOOD }

    result.onSuccess { r, _ -> assertThat(r).isEqualTo(TestResult.GOOD) }
      .onFailure { _, _ -> fail(RESULT_FAILURE) }
  }

  @Test
  fun `executing the policy function and failing with a handled exception type should return failure result indicating that exception type is one handled by this policy`() =
    runBlockingTest {
      val handledException = ArithmeticException()

      val result = Policy.asyncRetry<Unit>()
        .handle(ArithmeticException::class.java)
        .retry()
        .executeAndCapture { throw handledException }

      result.onFailureWithException { throwable, exceptionType, faultType, _ ->
        assertThat(throwable).isSameAs(handledException)
        assertThat(exceptionType).isSameAs(ExceptionType.HANDLED_BY_THIS_POLICY)
        assertThat(faultType).isSameAs(FaultType.EXCEPTION_HANDLED_BY_THIS_POLICY)
      }
        .onSuccess { _, _ -> fail(RESULT_SUCCESS) }
        .onFailureWithResult { _, _, _ -> fail(RESULT_FAILURE_WITH_RESULT) }
    }

  @Test
  fun `executing the policy function and failing with an unhandled exception type should return failure result indicating that exception type is unhandled by this policy`() =
    runBlockingTest {
      val unhandledException = Exception()

      val result = Policy.asyncRetry<Unit>()
        .handle(ArithmeticException::class.java)
        .retry()
        .executeAndCapture { throw unhandledException }

      result.onFailureWithException { throwable, exceptionType, faultType, _ ->
        assertThat(throwable).isSameAs(unhandledException)
        assertThat(exceptionType).isSameAs(ExceptionType.UNHANDLED)
        assertThat(faultType).isSameAs(FaultType.UNHANDLED_EXCEPTION)
      }
        .onSuccess { _, _ -> fail(RESULT_SUCCESS) }
        .onFailureWithResult { _, _, _ -> fail(RESULT_FAILURE_WITH_RESULT) }
    }

  @Test
  fun `executing the policy function and failing with a handled result should return failure result indicating that result is one handled by this policy`() =
    runBlockingTest {
      val handledResult = TestResult.FAULT

      val result = Policy.asyncRetry<TestResult>()
        .handleResult(handledResult)
        .retry()
        .executeAndCapture { handledResult }

      result.onFailureWithResult { testResult, faultType, _ ->
        assertThat(testResult).isSameAs(handledResult)
        assertThat(faultType).isSameAs(FaultType.RESULT_HANDLED_BY_THIS_POLICY)
      }
        .onSuccess { _, _ -> fail(RESULT_SUCCESS) }
        .onFailureWithException { _, _, _, _ -> fail(RESULT_FAILURE_WITH_EXCEPTION) }
    }

  @Test
  fun `executing the policy function and returning an unhandled result should return result not indicating any failure`() =
    runBlockingTest {
      val handledResult = TestResult.FAULT
      val unhandledResult = TestResult.GOOD

      val result = Policy.asyncRetry<TestResult>()
        .handleResult(handledResult)
        .retry()
        .executeAndCapture { unhandledResult }

      @Suppress("UsePropertyAccessSyntax")
      assertThat(result.isSuccess).isTrue()
      assertThat((result as PolicyResult.Success).result).isSameAs(unhandledResult)
    }
  //</editor-fold>

  //<editor-fold desc="context tests">
  @Test
  fun `executing the policy function should pass context to executed delegate`() = runBlockingTest {
    val operationKey = "SomeKey"
    val executionContext = Context(operationKey)
    var capturedContext: Context? = null

    Policy.asyncNoOp<Unit>()
      .execute(executionContext) { context -> capturedContext = context }

    assertThat(capturedContext?.operationKey).isEqualTo(operationKey)
  }

  @Test
  fun `executing and capturing the policy function should pass context to executed delegate`() = runBlockingTest {
    val operationKey = "SomeKey"
    val executionContext = Context(operationKey)
    var capturedContext: Context? = null

    Policy.asyncNoOp<Unit>()
      .executeAndCapture(executionContext) { context -> capturedContext = context }

    assertThat(capturedContext?.operationKey).isEqualTo(operationKey)
  }

  @Test
  fun `executing and capturing the policy function should pass context to policy result`() = runBlockingTest {
    val operationKey = "SomeKey";
    val executionContext = Context(operationKey)

    val result = Policy.asyncNoOp<Unit>()
      .executeAndCapture(executionContext) { _ -> }

    assertThat(result.context.operationKey).isEqualTo(operationKey)
  }
  //</editor-fold>
}
